pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/NoDictionaryOnHeapDictionaryJointRule.java

```diff
@@ -0,0 +1,245 @@
+import org.apache.pinot.controller.recommender.rules.io.params.NoDictionaryOnHeapDictionaryJointRuleParams;
+import org.apache.pinot.controller.recommender.rules.utils.FixedLenBitset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.NoDictionaryOnHeapDictionaryJointRule.*;
+import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.REALTIME;
+
+
+public class NoDictionaryOnHeapDictionaryJointRule extends AbstractRule {
+  private final Logger LOGGER = LoggerFactory.getLogger(NoDictionaryOnHeapDictionaryJointRule.class);
+  private final BrokerRequestOptimizer _brokerRequestOptimizer = new BrokerRequestOptimizer();
+  private final NoDictionaryOnHeapDictionaryJointRuleParams _params;
+
+  public NoDictionaryOnHeapDictionaryJointRule(InputManager inputManager, ConfigManager outputManager) {
+    super(inputManager, outputManager);
+    _params = inputManager.getNoDictionaryOnHeapDictionaryJointRuleParams();
+  }
+
+  @Override
+  public void run() {
+    LOGGER.info("Recommending no dictionary and on-heap dictionaries");
+
+    int numCols = _inputManager.getNumCols();
+    double[] filterGroupByWeights = new double[numCols];
+    double[] selectionWeights = new double[numCols];
+    AtomicDouble totalWeight = new AtomicDouble(0);
+
+    //**********No dictionary recommendation*******/
+    Set<String> noDictCols = new HashSet<>(_inputManager.getColNameToIntMap().keySet());
+
+    //Exclude cols with index
+    noDictCols.removeAll(_outputManager.getIndexConfig().getInvertedIndexColumns());
+    noDictCols.removeAll(_outputManager.getIndexConfig().getSortedColumn());
+    // TODO: Remove this after range index is implemented for no-dictionary
+    noDictCols.removeAll(_outputManager.getIndexConfig().getRangeIndexColumns());
+    LOGGER.debug("noDictCols {}", noDictCols);
+    //Find out columns used in filter&groupby and selection and corresponding frequencies
+    _inputManager.getQueryWeightMap().forEach((query, weight) -> {
+      parseQuery(query, weight, filterGroupByWeights, selectionWeights);
+      totalWeight.addAndGet(weight);
+    });
+
+    //Add dictionary on columns used in filter&groupby , with frequency > threshold
+    for (int i = 0; i < numCols; i++) {
+      double filterGroupByFreq = filterGroupByWeights[i] / totalWeight.get();
+      if (filterGroupByFreq > _params.THRESHOLD_MIN_FILTER_FREQ_DICTIONARY) {
+        noDictCols.remove(_inputManager.intToColName(i));
+      }
+    }
+
+    LOGGER.debug("filterGroupByWeights {}, selectionWeights{}, totalWeight{} ", filterGroupByWeights, selectionWeights,
+        totalWeight);
+    LOGGER.debug("noDictCols {}", noDictCols);
+
+    for (int i = 0; i < numCols; i++) {
+      // No dictionary on columns frequently used in selection
```

Comment:
Might want to add comment about the experiment done during the design phase -- for a column heavily used in selection only (not part of filter or group by), making it no dictionary reduces the latency by 20% (I guess) since we avoid the 2 lookups.

Reply:
It looks like we can improve the performance of this loop starting at line 94 by going over the remaining columns

Reply:
added